pysnooper/formatting.py

```diff
@@ -0,0 +1,172 @@
+import datetime as datetime_module
+import re
+
+from . import utils
+from .third_party import six
+
+
+class Event(object):
+    def __init__(self, frame, event, arg, depth):
+        self.frame = frame
+        self.event = event
+        self.arg = arg
+        self.depth = depth
+
+        self.variables = []
+        self.source = get_source_from_frame(self.frame)
+        self.line_no = self.frame.f_lineno
+
+        if self.event == 'call' and self.source_line.lstrip().startswith('@'):
+            # If a function decorator is found, skip lines until an actual
+            # function definition is found.
+            while True:
+                self.line_no += 1
+                try:
+                    if self.source_line.lstrip().startswith('def'):
+                        break
+                except IndexError:
+                    self.line_no = self.frame.lineno
+                    break
+
+    @property
+    def source_line(self):
+        return self.source[self.line_no - 1]
+
+    @property
+    def entries(self):
+        result = self.variables[:]
+        result.append(self)
+        if self.event == 'return':
+            result.append(ReturnValueEntry(
+                value=self.arg,
+            ))
+        return result
+
+
+class Entry(object):
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+
+class VariableEntry(Entry):
+    pass
+
+
+class ReturnValueEntry(Entry):
+    pass
+
+
+class DefaultFormatter(object):
+    datetime_format = '%H:%M:%S.%f'
+
+    def __init__(self, prefix):
+        self.prefix = six.text_type(prefix)
+        self.now_length = len(self.now_string())
+
+    def now_string(self):
+        return datetime_module.datetime.now().strftime(self.datetime_format)
+
+    def format(self, event):
+        indent = event.depth * u'    '
+        return ''.join([
+            (
+                    self.prefix
+                    + indent
+                    + self.format_entry(entry, event)
+                    + u'\n'
+            )
+            for entry in event.entries
+        ])
+
+    def format_entry(self, entry, event):
+        method = getattr(self, 'format_' + entry.__class__.__name__)
```

Comment:
Since this is a dynamic function finder perhaps we should catch the AttributeError and handle it somehow?

Reply:
It's not that dynamic, it's limited to the contents of that file. The user can't create new entry classes.

Reply:
Yeah I was talking more about the contributors not the users. If perhaps handling the exception with a more descriptive error could help future contributions that may require new entry types. But it is also fine having exceptions in development so it's not that major.